src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs

```diff
@@ -0,0 +1,299 @@
+            Host.CheckParam(windowSize >= 0, nameof(windowSize), "Must be non-negative.");
+            // REVIEW: Very bad design. This base class is responsible for reporting errors on
+            // the arguments, but the arguments themselves are not derived form any base class.
+            Host.CheckNonEmpty(inputColumnName, nameof(PercentileThresholdTransform.Arguments.Source));
+            Host.CheckNonEmpty(outputColumnName, nameof(PercentileThresholdTransform.Arguments.Source));
+
+            InputColumnName = inputColumnName;
+            OutputColumnName = outputColumnName;
+            OutputColumnType = outputColType;
+            InitialWindowSize = initialWindowSize;
+            WindowSize = windowSize;
+        }
+
+        protected SequentialTransformerBase(IHostEnvironment env, ModelLoadContext ctx, string name)
+        {
+            Contracts.CheckRef(env, nameof(env));
+            Host = env.Register(name);
+            Host.CheckValue(ctx, nameof(ctx));
+
+            // *** Binary format ***
+            // int: _windowSize
+            // int: _initialWindowSize
+            // int (string ID): _inputColumnName
+            // int (string ID): _outputColumnName
+            // ColumnType: _transform.Schema.GetColumnType(0)
+
+            var windowSize = ctx.Reader.ReadInt32();
+            Host.CheckDecode(windowSize >= 0);
+
+            var initialWindowSize = ctx.Reader.ReadInt32();
+            Host.CheckDecode(initialWindowSize >= 0);
+
+            var inputColumnName = ctx.LoadNonEmptyString();
+            var outputColumnName = ctx.LoadNonEmptyString();
+
+            InputColumnName = inputColumnName;
+            OutputColumnName = outputColumnName;
+            InitialWindowSize = initialWindowSize;
+            WindowSize = windowSize;
+
+            BinarySaver bs = new BinarySaver(Host, new BinarySaver.Arguments());
+            OutputColumnType = bs.LoadTypeDescriptionOrNull(ctx.Reader.BaseStream);
+        }
+
+        public virtual void Save(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+            Host.Assert(InitialWindowSize >= 0);
+            Host.Assert(WindowSize >= 0);
+
+            // *** Binary format ***
+            // int: _windowSize
+            // int: _initialWindowSize
+            // int (string ID): _inputColumnName
+            // int (string ID): _outputColumnName
+            // ColumnType: _transform.Schema.GetColumnType(0)
+
+            ctx.Writer.Write(WindowSize);
+            ctx.Writer.Write(InitialWindowSize);
+            ctx.SaveNonEmptyString(InputColumnName);
+            ctx.SaveNonEmptyString(OutputColumnName);
+
+            var bs = new BinarySaver(Host, new BinarySaver.Arguments());
+            bs.TryWriteTypeDescription(ctx.Writer.BaseStream, OutputColumnType, out int byteWritten);
+        }
+
+        public virtual Schema GetOutputSchema(Schema inputSchema)
```

Comment:
>virtual [](start = 15, length = 7)

if every override is not calling the base, do we need it present? I think it should be made abstract